contracts/authorities/SellPartyCardsAuthority.sol

```diff
@@ -0,0 +1,636 @@
+        party.increaseTotalVotingPower(votingPower);
+
+        return (votingPower, contribution);
+    }
+
+    function _batchContribute(
+        Party party,
+        uint256 saleId,
+        uint96[] memory contributions,
+        bytes calldata gateData
+    ) private returns (uint96[] memory votingPowers, uint96[] memory /* contributions */) {
+        SaleState memory state = _validateContribution(party, saleId, gateData);
+
+        uint256 numOfContributions = contributions.length;
+        uint96 totalValue;
+        uint96 totalVotingPower;
+        votingPowers = new uint96[](numOfContributions);
+        for (uint256 i; i < numOfContributions; ++i) {
+            uint96 contribution = contributions[i];
+            uint96 votingPower;
+
+            (votingPower, contributions[i], state.totalContributions) = _processContribution(
+                party,
+                saleId,
+                state,
+                contribution
+            );
+
+            votingPowers[i] = votingPower;
+
+            totalValue += contribution;
+            totalVotingPower += votingPower;
+        }
+
+        if (msg.value != totalValue) revert InvalidMessageValue();
+
+        // Transfer amount due to the Party. Revert if the transfer fails.
+        payable(address(party)).transferEth(address(this).balance);
+
+        party.increaseTotalVotingPower(totalVotingPower);
+
+        return (votingPowers, contributions);
+    }
+
+    /// @dev `totalContributions` is updated and returned for use in
+    ///      `batchContribute` and `batchContributeFor`.
+    function _processContribution(
+        Party party,
+        uint256 saleId,
+        SaleState memory state,
+        uint96 amount
+    ) private returns (uint96 votingPower, uint96 contribution, uint96 totalContributions) {
+        totalContributions = state.totalContributions;
+        uint96 maxTotalContributions = state.maxTotalContributions;
+
+        // Check sale is active.
+        if (
+            !_isSaleActive(
+                state.expiry,
+                totalContributions,
+                state.minContribution,
+                maxTotalContributions
+            )
+        ) {
+            revert SaleInactiveError();
+        }
+
+        // Check that the contribution amount is at or below the maximum.
+        uint96 maxContribution = state.maxContribution;
+        if (amount > maxContribution) {
+            revert OutOfBoundsContributionsError(amount, maxContribution);
+        }
+
+        uint96 newTotalContributions = totalContributions + amount;
+        if (newTotalContributions >= maxTotalContributions) {
+            // This occurs before refunding excess contribution to act as a
+            // reentrancy guard.
+            _saleStates[party][saleId]
+                .totalContributions = totalContributions = maxTotalContributions;
+
+            // Finalize the crowdfund.
+            emit Finalized(party, saleId);
```

Comment:
Wont emit finalized if `newTotalContributions >  maxTotalContributions - minContribution` but the the sale won't be active anymore.